src/swarm/neo/client/mixins/BatchRequestCore.d

```diff
@@ -0,0 +1,257 @@
+    resume(), and stop(). (Note that this interface is not provided in the
+    library purely in order to keep all API definitions for each request inside
+    a single module, not scattered around different modules in swarm.)
+
+    Params:
+        Request = type of request struct
+        IController = interface defining the controller API as presented to the
+            user. Assumed to contain methods suspend(), resume(), and stop() (in
+            addition to any other methods required by the request).
+        MessageType = type of enum defining the types of messages used by this
+            request's protocol. Expected to contain the members Suspend, Resume,
+            and Stop (in addition to any others required by the request).
+
+*******************************************************************************/
+
+public template BatchController ( Request, IController )
+{
+    /***************************************************************************
+
+        Request controller, accessible to the user via the client's `control()`
+        method.
+
+    ***************************************************************************/
+
+    public scope class Controller : IController
+    {
+        import ocean.core.Enforce;
+        import swarm.neo.client.mixins.RequestCore : ControllerBase;
+
+        /***********************************************************************
+
+            Base mixin.
+
+        ***********************************************************************/
+
+        mixin ControllerBase;
+
+        /***********************************************************************
+
+            Tells the nodes to stop sending data to this request.
+
+            Returns:
+                true because this controller function can always be used
+
+        ***********************************************************************/
+
+        public bool suspend ( )
+        {
+            Request.getContext(this.request_controller.context_blob)
+                .shared_working.suspendable_control.suspended = true;
+            return true;
+        }
+
+        /***********************************************************************
+
+            Tells the nodes to resume sending data to this request.
+
+            Returns:
+                true because this controller function can always be used
+
+        ***********************************************************************/
+
+        public bool resume ( )
+        {
+            Request.getContext(this.request_controller.context_blob)
+                .shared_working.suspendable_control.suspended = false;
+            this.request_controller.resumeSuspendedHandlers(
+                BatchRequestSharedWorkingData.Signal.Resume);
+            return true;
+        }
+
+        /***********************************************************************
+
+            Tells the nodes to cleanly end the request.
```

Comment:
Maybe also state that further data may be passed to the user, even after stop has been called.